pretyping/reductionops.ml

```diff
@@ -974,37 +956,39 @@ let rec whd_state_gen ?csts ~refold ~tactic_mode flags env sigma =
           else (* Looks for ReductionBehaviour *)
             match ReductionBehaviour.get (Globnames.ConstRef c) with
             | None -> whrec (Cst_stack.add_cst (mkConstU const) cst_l) (body, stack)
-            | Some (recargs, nargs, flags) ->
-              if (List.mem `ReductionNeverUnfold flags
-                  || (nargs > 0 && Stack.args_size stack < nargs))
-              then fold ()
-              else (* maybe unfolds *)
-              if List.mem `ReductionDontExposeCase flags then
-                let app_sk,sk = Stack.strip_app stack in
-                let (tm',sk'),cst_l' =
-                  whrec (Cst_stack.add_cst (mkConstU const) cst_l) (body, app_sk)
-                in
-                let rec is_case x = match EConstr.kind sigma x with
-                  | Lambda (_,_, x) | LetIn (_,_,_, x) | Cast (x, _,_) -> is_case x
-                  | App (hd, _) -> is_case hd
-                  | Case _ -> true
-                  | _ -> false in
-                if equal_stacks sigma (x, app_sk) (tm', sk')
-                || Stack.will_expose_iota sk'
-                || is_case tm'
-                then fold ()
-                else whrec cst_l' (tm', sk' @ sk)
-              else match recargs with
-                |[] -> (* if nargs has been specified *)
-                  (* CAUTION : the constant is NEVER refold
-                                          (even when it hides a (co)fix) *)
-                  whrec cst_l (body, stack)
-                |curr::remains -> match Stack.strip_n_app curr stack with
-                              | None -> fold ()
-                              | Some (bef,arg,s') ->
-                                whrec Cst_stack.empty
-                                        (arg,Stack.Cst(Stack.Cst_const (fst const, u'),curr,remains,bef,cst_l)::s')
-         end
+            | Some behavior ->
+              begin match behavior with
+              | ReductionBehaviour.NeverUnfold -> fold ()
+              | ReductionBehaviour.UnfoldWhen { nargs = Some n } when Stack.args_size stack < n ->
+                fold ()
+              | ReductionBehaviour.UnfoldWhen { recargs; dontexposematch } -> (* maybe unfolds *)
+                if dontexposematch then
+                  let app_sk,sk = Stack.strip_app stack in
```

Comment:
@gares isn't it suspicious that we ignore `recargs` in this branch?

Reply:
I think it fine, but I suggest you make 2 branches in the match, putting first the `dontexposematch = true` one.
Note that this case only inserts a stack frame, the recursive call eventually looks at recargs.